pkg/scheduler/framework/runtime/framework_test.go

```diff
@@ -338,6 +338,8 @@ var registry = func() Registry {
 	r.Register(pluginNotImplementingScore, newPluginNotImplementingScore)
 	r.Register(duplicatePluginName, newDuplicatePlugin)
 	r.Register(testPlugin, newTestPlugin)
+	r.Register(queueSortPlugin, newQueueSortPlugin)
+	r.Register(bindPlugin, newBindPlugin)
```

Comment:
Failed in my local environment, I don't know why prow ci passed.

Reply:
Could you share what the failures were? It's interesting that this didn't fail in CI

Reply:
command: 
```
go test -v --timeout 30s ./pkg/scheduler/framework/runtime -run ^TestNewFrameworkMultiPointExpansion$
````

output: 
```
=== RUN   TestNewFrameworkMultiPointExpansion/disable_and_enable_MultiPoint_plugins_with_'*'
    framework_test.go:761: Unexpected error, got PreFilterPlugin "no-op-queue-sort-plugin" does not exist, expect: 
--- FAIL: TestNewFrameworkMultiPointExpansion (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/plugin_expansion (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/disable_MultiPoint_plugin_at_some_extension_points (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/Multiple_MultiPoint_plugins (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/disable_MultiPoint_extension (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/Reorder_MultiPoint_plugins_(specified_extension_takes_precedence) (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/Override_MultiPoint_plugins_weights (0.00s)
    --- FAIL: TestNewFrameworkMultiPointExpansion/disable_and_enable_MultiPoint_plugins_with_'*' (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x1eeca9e]
```
for we haven't register these plugins. @damemi

Reply:
@kerthcet is that truncated output, or is there a fuller stack trace pointing to where the dereference happens?

Reply:
oh, sorry. full stack trace attached below:
```
=== RUN   TestNewFrameworkMultiPointExpansion/Reorder_MultiPoint_plugins_(specified_extension_takes_precedence)
I1203 10:32:23.008407   95878 framework.go:467] "MultiPoint plugin is explicitly re-configured; overriding" plugin="test-plugin"
I1203 10:32:23.008583   95878 framework.go:467] "MultiPoint plugin is explicitly re-configured; overriding" plugin="score-plugin-1"
=== RUN   TestNewFrameworkMultiPointExpansion/Override_MultiPoint_plugins_weights
I1203 10:32:23.008799   95878 framework.go:467] "MultiPoint plugin is explicitly re-configured; overriding" plugin="test-plugin"
I1203 10:32:23.008816   95878 framework.go:467] "MultiPoint plugin is explicitly re-configured; overriding" plugin="score-plugin-1"
=== RUN   TestNewFrameworkMultiPointExpansion/disable_and_enable_MultiPoint_plugins_with_'*'
    framework_test.go:761: Unexpected error, got PreFilterPlugin "no-op-queue-sort-plugin" does not exist, expect: 
--- FAIL: TestNewFrameworkMultiPointExpansion (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/plugin_expansion (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/disable_MultiPoint_plugin_at_some_extension_points (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/Multiple_MultiPoint_plugins (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/disable_MultiPoint_extension (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/Reorder_MultiPoint_plugins_(specified_extension_takes_precedence) (0.00s)
    --- PASS: TestNewFrameworkMultiPointExpansion/Override_MultiPoint_plugins_weights (0.00s)
    --- FAIL: TestNewFrameworkMultiPointExpansion/disable_and_enable_MultiPoint_plugins_with_'*' (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x1eeca9e]

goroutine 56 [running]:
testing.tRunner.func1.2({0x20a1e60, 0x2fe7ad0})
        /Users/kerthcet/sdk/go1.17/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /Users/kerthcet/sdk/go1.17/src/testing/testing.go:1212 +0x218
panic({0x20a1e60, 0x2fe7ad0})
        /Users/kerthcet/sdk/go1.17/src/runtime/panic.go:1038 +0x215
k8s.io/kubernetes/pkg/scheduler/framework/runtime.TestNewFrameworkMultiPointExpansion.func1(0xc000491520)
        /Users/kerthcet/Workspaces/kubernetes/pkg/scheduler/framework/runtime/framework_test.go:764 +0x19e
testing.tRunner(0xc000491520, 0xc000418b30)
        /Users/kerthcet/sdk/go1.17/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /Users/kerthcet/sdk/go1.17/src/testing/testing.go:1306 +0x35a
FAIL    k8s.io/kubernetes/pkg/scheduler/framework/runtime       0.779s
FAIL
```


we enabled three plugins [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework_test.go#L698-L700), but we haven't register them [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework_test.go#L333-L342) when we initialize a new Framework

Reply:
> The error comes from this line which makes it seem like fw is nil here when we try to call fw.ListPlugins().

It looks to me the error happens at [expandMultiPointPlugins()](https://github.com/kubernetes/kubernetes/blob/420c5308bbf13eeb93cfab3319d37e977e970c12/pkg/scheduler/framework/runtime/framework.go#L342) where the plugins (no-op-queue-sort-plugin, etc.) were enabled in multiPoint but didn't get registered.

> It's also still odd to me that this didn't fail when we were merging these changes.

This is also odd to me...

Reply:
@Huang-Wei from the stack trace it points to `ubernetes/pkg/scheduler/framework/runtime/framework_test.go:764` which is the cmp diff check for the expected vs received plugins. Where did you see that section?

To clarify, I'm sure the error is being caused somewhere around what you pointed out, but we have an additional problem if the test is not catching that error before this point

Reply:
> Where did you see that section?

The error does occur in cmp.Diff(). My point is that expandMultiPointPlugins() shouldn't expect an error, right? And the previous line loosens the error check too much:

```
if (err != nil && tc.wantErr == "") || (err == nil && tc.wantErr != "") || (err != nil && !strings.Contains(err.Error(), tc.wantErr)) {
```

In this case, `tc.wantErr` is "", so the last condition "!strings.Contains(err.Error(), tc.wantErr)" is always false.

Reply:
That's what I was trying to get at, I didn't have a chance to look at exactly what the check was but since the test wasn't failing here we have 2 problems:
* the multiPoint expansion fails without the plugins registered (I think that's fixed in this PR)
* the test doesn't report that error if it happens unexpectedly

So we need to fix up this error check still

Reply:
Sorry for the delayed response for I was on vacation last several days.
This was because of the [TestInitFrameworkWithScorePlugins](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework_test.go#L395-L442) function which will call [newFrameworkWithQueueSortAndBind](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework_test.go#L378-L393) to register these two plugins globally. 
Prow ci will run the unit tests with the who package but I used to run with a single testcase, that's the root cause I think.

Reply:
@kerthcet sounds good then. Could you still take a look at the error check as @Huang-Wei pointed out and try to fix that in the tests?

Reply:
sure, absolutely.

Reply:
I thought we have already catch all the errors: 
If we haven't registered these two plugins, `expandMultiPointPlugins` will throw an error and we'll caught it as below:
```
    framework_test.go:761: Unexpected error, got PreFilterPlugin "no-op-queue-sort-plugin" does not exist, expect: 
```
we can break up our [logic](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework_test.go#L760-L762) like this:
```
if err != nil {
	if tc.wantErr == "" || !strings.Contains((err.Error(), tc.wantErr)) {
		not pass
	}else{
		pass
	}
}else {
	if tc.wantErr == "" {
		pass
	}else {
		not pass
	}
}
```

I replace `t.Errorf` with `t.Fatalf` to prevent panic.

Reply:
@damemi any ideas?

Reply:
@kerthcet yeah, expanding the logic like that might make it a bit clearer what's being checked. it's definitely more verbose, but with all these cases to check it will be helpful to make sure we're covering everything